File: apps/ffmpeg/src/ffmpeg.erl
%%% @author     Max Lapshin <max@maxidoors.ru> [http://erlyvideo.org]
%%% @copyright  2010-2012 Max Lapshin
%%% @doc        multibitrate packetizer
%%% @reference  See <a href="http://erlyvideo.org" target="_top">http://erlyvideo.org</a> for more information
%%% @end
%%%
%%%
%%% This file is part of erlyvideo.
%%% 
%%% erlyvideo is free software: you can redistribute it and/or modify
%%% it under the terms of the GNU General Public License as published by
%%% the Free Software Foundation, either version 3 of the License, or
%%% (at your option) any later version.
%%%
%%% erlyvideo is distributed in the hope that it will be useful,
%%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%%% GNU General Public License for more details.
%%%
%%% You should have received a copy of the GNU General Public License
%%% along with erlmedia.  If not, see <http://www.gnu.org/licenses/>.
%%%
%%%---------------------------------------------------------------------------------------
-module(ffmpeg).
-include_lib("eunit/include/eunit.hrl").
-include_lib("erlmedia/include/video_frame.hrl").
-include_lib("erlmedia/include/media_info.hrl").
-include("log.hrl").

-type(output_option() :: 
  {width,Width::non_neg_integer()}
  |{height,Height::non_neg_integer()}
  |{bitrate,Bitrate::non_neg_integer()}
  |{FFkey::string(),FFvalue::string()}
  ).

-record(init_output, {
  content = undefined :: frame_content(),
  codec = undefined :: frame_codec(),
  track_id = undefined :: non_neg_integer(),
  options = [] :: [output_option()]
}).

-record(init_input, {
  content = undefined :: frame_content(),
  codec = undefined :: frame_codec(),
  config = <<>> :: binary()
}).

-export([start_link/0]).
-export([init/1, handle_call/3, handle_info/2, terminate/2]).

-export([start_worker/0, start_worker/1, send/2, fetch/1, fetch/2, close/1]).
-export([init_decoder/2, send_frame/2]).
-export([init_encoder/2]).

start_link() ->
  gen_server:start_link(?MODULE, [], []).


start_worker() ->
  start_worker(filename:join(code:lib_dir(ffmpeg, priv), "flussonic_ffmpeg")).

start_worker(Path) ->
  case string:tokens(Path, ":") of
    [Host, Port] ->
      {ok, Socket} = gen_tcp:connect(Host, list_to_integer(Port), [binary,{packet,4},{send_timeout,3000},{active,false}]),
      {socket, Socket};
    _ ->
      Port = erlang:open_port({spawn_executable, Path}, [{packet,4},{arg0, "flussonic_ffmpeg"},binary,exit_status]),
      {program, Port}
  end.


close({socket, Sock}) ->
  gen_tcp:close(Sock);

close({program, Port}) ->
  erlang:port_close(Port).


send({socket, Sock}, Term) ->
  gen_tcp:send(Sock, erlang:term_to_binary(Term));

send({program, Port}, Term) ->
  erlang:port_command(Port, erlang:term_to_binary(Term)),
  ok.


send_frame(_Worker, #video_frame{flavor = config}) ->
  {error, config_forbidden};

% send_frame(Worker, #video_frame{codec = h264, flavor = Flavor, body = Body} = Frame) ->
%   send(Worker, Frame)

send_frame(Worker, #video_frame{} = Frame) ->
  send(Worker, Frame),
  Reply = fetch(Worker),
  Reply.



init_decoder(Port, #media_info{streams = Streams}) ->
  [begin
    send(Port, #init_input{content = Content, codec = Codec, config = Config}),
    case fetch(Port) of
      ready -> ok;
      Else -> error({init_input,Codec,TrackId,Else})
    end
  end || #stream_info{content = Content, codec = Codec, config = Config, track_id = TrackId} <- Streams],
  ok.


ev_to_av(h264) -> libx264;
ev_to_av(aac) -> libfaac;
ev_to_av(Codec) -> Codec.

av_to_ev(libx264) -> h264;
av_to_ev(libfaac) -> aac;
av_to_ev(Codec) -> Codec.

init_encoder(Port, #media_info{streams = Streams}) ->
  Streams1 = lists:map(fun
    (#stream_info{content = video, params = #video_params{width = W, height = H}, codec = C, bitrate = B, options = O} = S) ->
      S#stream_info{options = [{"preset","fast"},{width,W},{height,H},{bitrate,B}|O], codec = ev_to_av(C)};
    (#stream_info{content = audio, codec = C, bitrate = B, options = O, params = #audio_params{sample_rate = SR, channels = Ch}} = S) ->
      S#stream_info{codec = ev_to_av(C), options = [{bitrate,B},{sample_rate,SR},{channels,Ch}|O]}
  end, Streams),
  Configs = lists:flatmap(fun(#stream_info{content = Content, codec = Codec, options = Options, track_id = TrackId}) ->
    % ?debugFmt("init_encoder: ~p", [{Content, Codec, TrackId, Options}]),
    send(Port, #init_output{content = Content, codec = Codec, track_id = TrackId, options = Options}),
    case fetch(Port) of
      ready -> [];
      #video_frame{} = F -> [F];
      Else -> error({init_output,Codec,TrackId,Else})
    end
  end, Streams1),
  {ok, Configs}.


fetch(Port) ->
  fetch(Port, 2000).

fetch({socket, Sock}, Timeout) ->
  case gen_tcp:recv(Sock, 0, Timeout) of
    {ok, Data} -> transform_reply(erlang:binary_to_term(Data));
    {error, closed} -> closed;
    {error, _} = Error -> Error
  end;

fetch({program, Port}, Timeout) ->
  receive
    {Port, {exit_status, 0}} -> closed;
    {Port, {exit_status, Code}} -> {exit, Code};
    {Port, {data, Data}} -> transform_reply(erlang:binary_to_term(Data));
    {Port, Data} -> {ok, Data}
  after
    Timeout -> {error, timeout}
  end.


transform_reply(#video_frame{codec = Codec} = Frame) ->
  transform_frame(Frame#video_frame{codec = av_to_ev(Codec)});
transform_reply(A) ->
  A.


transform_frame(#video_frame{codec = h264, body = Body, flavor = config} = Reply) ->
  NALs = [NAL || NAL <- binary:split(Body, [<<1:32>>, <<1:24>>], [global]), size(NAL) > 0],
  H264 = lists:foldl(fun(NAL, H_) -> {H1_,_} = h264:decode_nal(NAL, H_), H1_ end, h264:init(), NALs),
  % ?debugFmt("<  config ~p", [Reply#video_frame.dts]),
  Reply#video_frame{body = h264:decoder_config(H264)};
transform_frame(#video_frame{codec = h264, body = Annexb} = Reply) ->
  NALs = [NAL || NAL <- binary:split(Annexb, [<<1:32>>, <<1:24>>], [global]), size(NAL) > 0],
  Body = iolist_to_binary([[<<(size(NAL)):32>>, NAL] || NAL <- NALs]),
  % ?debugFmt("<~8s ~p ~p ~p", [Reply#video_frame.flavor, Reply#video_frame.dts, Reply#video_frame.pts,
  %   [{h264:type(NAL), size(NAL)} || NAL <- NALs]       ]),
  Reply#video_frame{body = Body};
transform_frame(#video_frame{} = Frame) ->
  Frame.


-record(ffmpeg, {
  port
}).

init([]) ->
  Port = start_worker(),
  self() ! init,
  {ok, #ffmpeg{port = Port}}.


handle_info(init, #ffmpeg{port = _Port} = FFmpeg) ->
  {noreply, FFmpeg};

handle_info(Info, State) ->
  {stop, {error, {unknown_info, Info}}, State}.


handle_call(Call, _From, State) ->
  {stop, {error, {unknown_call, Call}}, State}.


terminate(_,_) ->
  ok.

File: apps/ffmpeg/c_src/ffmpeg.c
// @author     Max Lapshin <max@maxidoors.ru> [http://erlyvideo.org]
// @copyright  2010-2012 Max Lapshin
// @doc        multibitrate packetizer
// @reference  See <a href="http://erlyvideo.org" target="_top">http://erlyvideo.org</a> for more information
// @end
//
//
// This file is part of erlyvideo.
// 
// erlyvideo is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// erlyvideo is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with erlmedia.  If not, see <http://www.gnu.org/licenses/>.
//
//---------------------------------------------------------------------------------------

#include <ei.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

#include <libavcodec/avcodec.h>
#include "reader.h"


typedef struct Track {
  AVCodec *codec;
  AVCodecContext *ctx;
  int track_id;
  enum AVMediaType content;
  enum AVCodecID codec_id;
  int width;
  int height;
} Track;

#define MAX_OUTPUT_TRACKS 4


Track input_audio;
Track input_video;
Track output_audio[MAX_OUTPUT_TRACKS];
int out_audio_count = 0;
Track output_video[MAX_OUTPUT_TRACKS];
int out_video_count = 0;

extern int out_fd;
extern int in_fd;

void loop();
void pong(void);
void debug_loop(int argc, char *argv[], void (*loop)(void));
void reply_atom(char *a);
void reply_avframe(AVPacket *pkt, AVCodec *codec);
void error(const char *fmt, ...);
ssize_t read1(int fd, void *buf, ssize_t len);


int main(int argc, char *argv[]) {
  avcodec_register_all();
  av_log_set_level(AV_LOG_ERROR);

  bzero(&input_audio, sizeof(Track));
  bzero(&input_video, sizeof(Track));
  bzero(&output_audio, sizeof(output_audio));
  bzero(&output_video, sizeof(output_video));

  argc--;
  argv++;
  if(argc >= 2  && !strcmp(argv[0], "-d")) {
    debug_loop(argc, argv, loop);
  } else {
    loop();
  }
}


void loop() {
  int64_t dts_shift = AV_NOPTS_VALUE;

  uint32_t buf_size = 10240;
  char *buf = (char *)malloc(buf_size);
  while(1) {
    uint32_t len;
    int idx = 0;
    int read_bytes = 0;
    if((read_bytes = read1(in_fd, &len, 4)) != 4) {
      if(read_bytes == 0) {
        _exit(0);
      }
      error("Can't read input length: %d", read_bytes);
    }
    len = ntohl(len);
    if(len > buf_size) {
      buf_size = len;
      free(buf);
      buf = (char *)malloc(buf_size);
    }

    if((read_bytes = read1(in_fd, buf, len)) != len) error("Can't read %d bytes from input: %d", len, read_bytes);
    int version = 0;
    ei_decode_version(buf, &idx, &version);
    int command_idx = idx;

    int arity = 0;
    if(ei_decode_tuple_header(buf, &idx, &arity) == -1) error("must pass tuple");


    int t = 0;
    int size = 0;
    ei_get_type(buf, &idx, &t, &size);
    if(t != ERL_ATOM_EXT) error("first element must be atom");
    char command[MAXATOMLEN+1];
    ei_decode_atom(buf, &idx, command); arity--;


    if(!strcmp(command, "ping")) {
      pong();
      continue;
    }
    if(!strcmp(command, "exit")) {
      return;
    }
    if(!strcmp(command, "init_input")) {
      if(arity != 3) error("Must provide 3 arguments to init_input command");
      char content[1024];
      char codec[1024];
      if(ei_decode_atom(buf, &idx, content) == -1) error("Must provide content as an atom");
      if(ei_decode_atom(buf, &idx, codec) == -1) error("Must provide codec as an atom");

      int decoder_config_len = 0;
      ei_get_type(buf, &idx, &t, &decoder_config_len);
      if(t != ERL_BINARY_EXT) error("decoder config must be a binary");
      uint8_t *decoder_config = av_mallocz(decoder_config_len + FF_INPUT_BUFFER_PADDING_SIZE);
      long bin_len = 0;
      ei_decode_binary(buf, &idx, decoder_config, &bin_len);

      Track *t = NULL;
      if(!strcmp(content, "video")) {
        t = &input_video;
      } else if(!strcmp(content, "audio")) {
        t = &input_audio;
      } else {
        error("Unknown media content: '%s'", content);
      }
      if(t->codec) error("Double initialization of media '%s'", content);

      t->codec = avcodec_find_decoder_by_name(codec);
      t->ctx = avcodec_alloc_context3(t->codec);
      if(!t->codec || !t->ctx) 
        error("Unknown %s decoder '%s'", content, codec);
      t->ctx->time_base = (AVRational){1, 90};
      t->ctx->extradata_size = decoder_config_len;
      t->ctx->extradata = decoder_config;
      if(avcodec_open2(t->ctx, t->codec, NULL) < 0) 
        error("failed to allocate %s decoder", content);

      reply_atom("ready");
      continue;
    }

    if(!strcmp(command, "init_output")) {
      if(arity != 4) error("Must provide 4 arguments to init_output command");
      char content[1024];
      char codec[1024];
      if(ei_decode_atom(buf, &idx, content) == -1) error("Must provide content as an atom");
      if(ei_decode_atom(buf, &idx, codec) == -1) error("Must provide codec as an atom");

      long track_id = -1;
      if(ei_decode_long(buf, &idx, &track_id) == -1) error("track_id must be integer");
      if(track_id < 1 || track_id > MAX_OUTPUT_TRACKS+1) error("track_id must be from 1 to %d", MAX_OUTPUT_TRACKS+1);
      track_id--;

      Track *t = NULL;
      if(!strcmp(content, "audio")) {
        t = &output_audio[out_audio_count++];
      } else if(!strcmp(content, "video")) {
        t = &output_video[out_video_count++];
      } else {
        error("invalid_content '%s'", content);
      }
      t->track_id = track_id;

      t->codec = avcodec_find_encoder_by_name(codec);
      t->ctx = avcodec_alloc_context3(t->codec);
      if(!t->codec || !t->ctx) error("Unknown encoder '%s'", codec);

      AVCodecContext* ctx = t->ctx;
      AVDictionary *opts = NULL;


      int options_count = 0;
      if(ei_decode_list_header(buf, &idx, &options_count) < 0) error("options must be a proplist");
      while(options_count > 0) {
        int arity1 = 0;

        int t,s;
        ei_get_type(buf, &idx, &t, &s);
        if(t == ERL_NIL_EXT) {
          ei_skip_term(buf, &idx);
          break;
        }

        if(ei_decode_tuple_header(buf, &idx, &arity1) < 0) error("options must be a proper proplist");
        if(arity1 != 2) error("tuples in options proplist must be arity 2");

        char key[MAXATOMLEN];
        if(ei_decode_atom(buf, &idx, key) == 0) {

          if(!strcmp(key, "width")) {
            long w = 0;
            if(ei_decode_long(buf, &idx, &w) < 0) error("width must be integer");
            ctx->width = w;
            continue;
          }

          if(!strcmp(key, "height")) {
            long h = 0;
            if(ei_decode_long(buf, &idx, &h) < 0) error("height must be integer");
            ctx->height = h;
            continue;
          }

          if(!strcmp(key, "bitrate")) {
            long b = 0;
            if(ei_decode_long(buf, &idx, &b) < 0) error("bitrate must be integer");
            ctx->bit_rate = b;
            continue;
          }

          if(!strcmp(key, "sample_rate")) {
            long sr = 0;
            if(ei_decode_long(buf, &idx, &sr) < 0) error("sample_rate must be integer");
            ctx->sample_rate = sr;
            continue;
          }

          if(!strcmp(key, "channels")) {
            long ch = 0;
            if(ei_decode_long(buf, &idx, &ch) < 0) error("channels must be integer");
            ctx->channels = ch;
            continue;
          }

          fprintf(stderr, "Unknown key: '%s'\r\n", key);
          ei_skip_term(buf, &idx);
          continue;
        } else if(ei_decode_string(buf, &idx, key) == 0) {
          char value[MAXATOMLEN];
          if(ei_decode_string(buf, &idx, value) < 0) error("key-value must be strings");
          av_dict_set(&opts, key, value, 0);
        } else {
          error("Invalid options proplist");
        }
      }

      if(!strcmp(content, "video")) {
        ctx->pix_fmt = AV_PIX_FMT_YUV420P;
      }
      if(!strcmp(content, "audio")) {
        ctx->sample_fmt = AV_SAMPLE_FMT_S16;
        ctx->profile = FF_PROFILE_AAC_MAIN;
      }
      ctx->flags |= CODEC_FLAG_GLOBAL_HEADER;
      ctx->time_base = (AVRational){1,90};

      if(avcodec_open2(ctx, t->codec, &opts) < 0) error("failed to allocate video encoder");

      AVPacket config;
      config.dts = config.pts = 0;
      config.flags = CODEC_FLAG_GLOBAL_HEADER;
      config.data = ctx->extradata;
      config.size = ctx->extradata_size;
      reply_avframe(&config, t->codec);      
      continue;
    }

    if(!strcmp(command, "video_frame")) {
      idx = command_idx;
      struct video_frame *fr = read_video_frame(buf, &idx);

      AVPacket packet;
      av_new_packet(&packet, fr->body.size);
      memcpy(packet.data, fr->body.data, fr->body.size);
      packet.size = fr->body.size;
      packet.dts = fr->dts*90;
      packet.pts = fr->pts*90;
      packet.stream_index = fr->track_id;

      // if(packet_size != pkt_size) error("internal error in reading frame body");

      if(fr->content == frame_content_audio) {
        if(!input_audio.ctx) error("input audio uninitialized");

        AVFrame *decoded_frame = avcodec_alloc_frame();
        int got_output = 0;
        int ret = avcodec_decode_audio4(input_audio.ctx, decoded_frame, &got_output, &packet);
        if(got_output) {
          reply_atom("ok");
        } else {
          error("Got: %d, %d\r\n", ret, got_output);
        }
        free(fr);
        continue;
      }

      if(fr->content == frame_content_video) {
        if(!input_video.ctx) error("input video uninitialized");
        AVFrame *decoded_frame = avcodec_alloc_frame();
        int could_decode = 0;
        int ret = avcodec_decode_video2(input_video.ctx, decoded_frame, &could_decode, &packet);
        if(ret < 0) {
          error("failed to decode video");
        }
        if(could_decode) {
          decoded_frame->pts = av_frame_get_best_effort_timestamp(decoded_frame);
          int sent_config = 0;

          AVPacket pkt;
          av_init_packet(&pkt);
          pkt.data = NULL;
          pkt.size = 0;

          int could_encode = 0;

          if(out_video_count <= 0) error("trying to transcode uninitialized video");
          if(avcodec_encode_video2(output_video[0].ctx, &pkt, decoded_frame, &could_encode) != 0) 
            error("Failed to encode h264");

          if(could_encode) {
            if(dts_shift == AV_NOPTS_VALUE) {
              dts_shift = -pkt.dts;
            }
            pkt.dts += dts_shift;
            reply_avframe(&pkt, output_video[0].codec);
          } else if(!sent_config) {
            reply_atom("ok");
          }
          free(fr);
          continue;
        } else {
          reply_atom("ok");
          free(fr);
          continue;
        }
      }

      error("Unknown content");
    }

    // AVCodecContext
    // AVPacket
    // AVFrame



    char *s = (char *)malloc(1024);
    ei_s_print_term(&s, buf, &command_idx);
    error("Unknown command: %s", s);
  }
}


void pong(void) {
  reply_atom("pong");
}

